File: java/reflect/src/reflect/Param.java
//
// Copyright (c) 2006, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   4 Jan 06  Brian Frank  Creation
//
package reflect;

import fan.sys.*;
import fanx.fcode.*;
import fanx.main.*;


/**
 * Param represents one parameter definition of a Func (or Method).
 */
public class Param
  extends FanObj
{

//////////////////////////////////////////////////////////////////////////
// Constructor
//////////////////////////////////////////////////////////////////////////

  public Param(String name, Type type, int mask)
  {
    this.name = name;
    this.type = type;
    this.mask = mask;
  }

//////////////////////////////////////////////////////////////////////////
// Methods
//////////////////////////////////////////////////////////////////////////
  private static Type typeof;

  public Type typeof() { 
	  if (typeof == null) {
		  typeof = Sys.findType("reflect::Param");
	  }
	  return typeof;
  }

  public String name()  { return name; }
  public Type type()   { return type; }
  public boolean hasDefault() { return (mask & HAS_DEFAULT) != 0; }

  public String toStr() { return type + " " + name; }

//////////////////////////////////////////////////////////////////////////
// Fields
//////////////////////////////////////////////////////////////////////////

  public static final int HAS_DEFAULT   = 0x01;  // is a default value provided

  final String name;
  final Type type;
  final int mask;

}

File: src/std/js/Param.js
//
// Copyright (c) 2009, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   18 May 09  Andy Frank  Creation
//   20 May 09  Andy Frank  Refactor to new OO model
//

/**
 * Param.
 */
fan.std.Param = fan.sys.Obj.$extend(fan.sys.Obj);

//////////////////////////////////////////////////////////////////////////
// Constructor
//////////////////////////////////////////////////////////////////////////

fan.std.Param.prototype.$ctor = function(name, type, hasDefault)
{
  this.m_name = name;
  this.m_type = (type instanceof fan.std.Type) ? type : fan.std.Type.find(type);
  this.m_hasDefault = hasDefault;
}

//////////////////////////////////////////////////////////////////////////
// Methods
//////////////////////////////////////////////////////////////////////////

fan.std.Param.prototype.$name = function() { return this.m_name; }
fan.std.Param.prototype.type = function() { return this.m_type; }
fan.std.Param.prototype.hasDefault = function() { return this.m_hasDefault; }
fan.std.Param.prototype.$typeof = function() { return fan.std.Param.$type; }
fan.std.Param.prototype.toStr = function() { return this.m_type.toStr() + " " + this.m_name; }
